### ISSUE TITLE
Config files for reproducing experiments

### ISSUE THREAD

**Issue by clblalock** [ngmarchant/dblink#4]
Hi Neil,

We've successfully run the dblink on our system and the evaluation metrics for RLdata10000 don't match those in your paper. 

Can you provide the config files for RLdata10000 that you used in your paper? We were testing the config file you included in the examples on the repo. We're also planning to test the ABSEmployee and SHIW0810 data that you provided. Can you provide config files for those as well?

How do you want us to provide the results to you? We can zip the output and post it here if that doesn't violate data any of your data agreements.

Thanks,
Casey

**Comment by resteorts** [comment]
@ngmarchant: It appears that the config for RLdata10000 may not match the new results in the paper. Could you look into this when you have a moment.

On Sep 19, 2019, at 9:07 AM, clblalock <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi Neil,

We've successfully run the dblink on our system and the evaluation metrics for RLdata10000 don't match those in your paper.

Can you provide the config files for RLdata10000 that you used in your paper? We were testing the config file you included in the examples on the repo. We're also planning to test the ABSEmployee and SHIW0810 data that you provided. Can you provide config files for those as well?

How do you want us to provide the results to you? We can zip the output and post it here if that doesn't violate data any of your data agreements.

Thanks,
Casey

**Comment by ngmarchant** [comment]
> We've successfully run the dblink on our system and the evaluation metrics for RLdata10000 don't match those in your paper.
> 
> Can you provide the config files for RLdata10000 that you used in your paper? We were testing the config file you included in the examples on the repo. 

I'm planning to create a separate repository with details/scripts for reproducing the experiments in the paper. The `RLdata10000.conf` file in this repo is really just for testing purposes. In practice, you need to use a longer burn-in period and draw more samples. The need for a longer burn-in period becomes apparent from looking at the diagnostics. You should see that the summary statistics have not converged after only 100 iterations. You can explore these by loading the `diagnostics.csv` in R or Excel, for example.

To reproduce the RLdata10000 experiment in our paper, you can use the following sample step
```
{name : "sample", parameters : {
            sampleSize : 10000,
            burninInterval : 0,
            thinningInterval : 10,
            resume : false,
            sampler : "PCG-I"
        }}
```
and the following evaluate step
```
{name : "evaluate", parameters : {
            lowerIterationCutoff : 10000,
            metrics : ["pairwise", "cluster"], 
            useExistingSMPC : false
        }}
```
The `lowerIterationCutoff` parameter above acts like a burnin, discarding any samples before iteration 10000.

> We're also planning to test the ABSEmployee and SHIW0810 data that you provided. Can you provide config files for those as well?

I've uploaded these [conf-files.zip](https://github.com/ngmarchant/dblink/files/3633701/conf-files.zip). Note that you'll have to update the path-related parameters before running.

> How do you want us to provide the results to you? We can zip the output and post it here if that doesn't violate data any of your data agreements.

If there's some kind of discrepancy or issue that you're concerned about, you can provide details here. Depending on the problem, you might include the log file, diagnostics, or evaluation measures. You won't be able to upload the full output, as the posterior samples can become quite large (10s of GB).

**Comment by YathishK** [comment]
RLdata10000.csv file has total of 10,000 records. Please could you confirm that sampleSize : 10000
in config file.
Thank you,
Yathish

**Comment by resteorts** [comment]
The total number of record is 10000. The total of duplicates is 10 percent. For more documentation, please see the Record Linkage package in R for this data set. Please let me know if I misunderstood your question.

Sent from my iPhone

On Sep 20, 2019, at 14:29, YathishK <notifications@github.com<mailto:notifications@github.com>> wrote:


RLdata10000.csv file has total of 10,000 records. Please could you confirm that sampleSize : 10000
in config file.
Thank you,
Yathish

**Comment by ngmarchant** [comment]
@YathishK Please refer to the [documentation](https://github.com/ngmarchant/dblink/blob/master/docs/configuration.md#sample-step) for information about the parameters, in particular the `sampleSize` parameter.

**Comment by YathishK** [comment]
Thank you. Attached here the configuration file that we are using. 
[RLdata10000.conf.txt](https://github.com/cleanzr/dblink/files/3648867/RLdata10000.conf.txt)

**Comment by resteorts** [comment]
@YathishK, please post the issues for RLdata10000 (and only these) here please so that these can be investigated. Please do not open a new issue. Thank you so much for your help with keeping one issue open at a time.

**Comment by resteorts** [comment]
@YathishK 

I have just reviewed the configuration file that you are running and it does not match the one that is provided on the repository. I would suggest that you re-download the configuration file for RLdata10000 and try re-running this please and report back. (If you compare the files, you will notice that the one that you posted is not the same one as one the repository). Please report back.

**Comment by YathishK** [comment]
I did compare both the config files using diff on linux 
**$ diff ~/cvap/RLdata10000.conf.txt ~/cvap/RLdata10000_from_examples.conf > ~/diff.txt**

The only difference is in sampleSize and  lowerIterationCutoff,, both fields increased from 100 to 10000 as Neil suggested  above. If you see any other changes , please let us know.

cat ~/diff.txt
63c63
<             sampleSize : 10000,
---
>             sampleSize : 100,
74c74
<             lowerIterationCutoff : 10000,
---
>             lowerIterationCutoff : 100,

**Comment by ngmarchant** [comment]
> Thank you. Attached here the configuration file that we are using.
> [RLdata10000.conf.txt](https://github.com/cleanzr/dblink/files/3648867/RLdata10000.conf.txt)

The config file looks fine, although you may need to update the path-related parameters `outputPath`, `checkpointPath` and `data.path`. If running in cluster mode, these should all be HDFS paths, not paths on the local filesystem.

> I have just reviewed the configuration file that you are running and it does not match the one that is provided on the repository. 

As mentioned above, the config file in the `examples` directory does not replicate the RLdata10000 experiment from the paper. 
@YathishK: just to clarify, the `sampleSize` and `lowerIterationCutoff` parameters are set correctly for reproducing our results.

**Comment by YathishK** [comment]
The file paths are correct , _outputPath_, _checkpointPath_ and _data.path_ all pointing to HDFS paths. We created file system on HDFS such a way that we don't need to make any changes to config file. Also .config file and data file copied to HDFS  file system _examples_ folder.

**Comment by ngmarchant** [comment]
> The file paths are correct , _outputPath_, _checkpointPath_ and _data.path_ all pointing to HDFS paths. We created file system on HDFS such a way that we don't need to make any changes to config file. Also .config file and data file copied to HDFS file system _examples_ folder.

This seems like an unusual setup. Can you describe it in more detail, or provide a reference? Based on the log files you uploaded in Issue #5, it seems Spark is interpreting the storage as local. I suspect this is preventing checkpointing from working correctly. This is more likely to be an issue for longer Markov chains, where the RDD lineage becomes very long.

My advice is to re-run the RLdata10000 experiment using the HDFS protocol for the storage locations.

**Comment by YathishK** [comment]
_My advice is to re-run the RLdata10000 experiment using the HDFS protocol for the storage locations._

Sure thing, I will check the file paths one more time and re-run the job first thing in the morning with 10K sample size.

I would like to mention that we successfully ran dblink on  RLdata10000  data with sample size of 100 (took 20 minutes)and sample size of 1000 (took 80 minutes) ,  Sample Size of 1000 results matched with the results in published paper.

In the mean time, could you please answer below questions  (for sample size of 10000).

1) Did you ever run dblink application with samplesize of 10000 ? 
2) What is the command you used to run with samplesize of 10000? 
3) What is your cluster size ?
4) How much time the job took to run sample size of 10000? 
5) What are your results for sample size of 10000?

**Comment by ngmarchant** [comment]
> Sure thing, I will check the file paths one more time and re-run the job first thing in the morning with 10K sample size.

Just to clarify, the paths should begin with the HDFS scheme, like `hdfs://path/`

> I would like to mention that we successfully ran dblink on RLdata10000 data with sample size of 100 (took 20 minutes)and sample size of 1000 (took 80 minutes) , Sample Size of 1000 results matched with the results in published paper.

This may have been luck. For longer chains, the risk of an error occurring on one of the executors is higher, and checkpointing becomes more important.

> In the mean time, could you please answer below questions (for sample size of 10000).
> 
> 1. Did you ever run dblink application with samplesize of 10000 ?

Yes. I think the largest sampleSize we've tested is 20000.

> 2. What is the command you used to run with samplesize of 10000?

For the experiments presented in the main paper, we ran Spark in local mode on a server. The spark-submit command is the same as the one in the guide.

When running in cluster mode on Amazon EMR, we use
```bash
$ spark-submit --deploy-mode cluster --class com.github.ngmarchant.dblink.Run "s3://dblink/dblink-assembly-0.1.jar" "s3://dblink/RLdata10000-experiment/RLdata10000.conf"
```

> 3. What is your cluster size ?

For the experiments run on our server, we set the number of threads equal to the number of partitions + 1. Further details are provided in the paper. For RLdata10000, we used 3 worker threads. For the largest data set (ABSEmployee) we used 65 worker threads.

For the experiments run on Amazon EMR, we provide some basic details in the paper. To summarize:
* Master node: an m4.large instance (4 vCore, 8 GiB memory, 32 GiB EBS storage)
* Slave nodes: m5.xlarge instances (4 vCore, 16 GiB memory, 32 GiB EBS storage)
We set the number of slave nodes so that we have one vCore per partition. So for RLdata10000, we used a master node + a single slave node.

> 4. How much time the job took to run sample size of 10000?

About 5 hrs for RLdata10000 on our local server. About 7 hrs on Amazon EMR.

> 5. What are your results for sample size of 10000?

These are the results reported in the paper.

**Comment by YathishK** [comment]
Again I ran the dblink application yesterday , started the job around 11.12 am and completed by 7.04 in the evening. It ran good.  Below are complete results, 
Today I will run again with sample size of 10K  by hard coding the HDFS paths in .config file as mentioned above.

dblink application run on RLdata10000  Results 

Sample size = _10000_  -  Application took around 8 hours to complete – Successful run on Sep 24th

Pairwise metrics           
 -------------------------------------
  Precision:       0.633015873015873
  Recall:          0.997
  F1-score:        0.774368932038835

 Cluster metrics            
 -------------------------------------
  Adj. Rand index: 0.7743634110575704

Sample size = _1000_  -  Application took around 80 minutes to complete – Successful run on Sep 18th  (Same results as in paper)

Pairwise metrics           
 -------------------------------------
  Precision:       0.6202134337727558
  Recall:          0.988
  F1-score:        0.762051677593521

 Cluster metrics            
 -------------------------------------
  Adj. Rand index: 0.7620458295579883

Sample size = _100_ -  Application took around 20 minutes to complete – Successful run on Sep 17th 

Pairwise metrics           
 -------------------------------------
  Precision:       0.6114570361145704
  Recall:          0.982
  F1-score:        0.7536454336147352

 Cluster metrics            
 -------------------------------------
  Adj. Rand index: 0.7536393600190746

**Comment by resteorts** [comment]
@YathishK Thank you so much for the update. We look forward to hearing back about the re-run later today. If things are not successful, please include the log files. If things are successful, please do let us know if changing the paths fixed the issue and also include the summary statistic information regarding evaluations and that you are able to reproduce results in the paper. Thank you again!

**Comment by YathishK** [comment]
dblink application completed successfully (10K sample size) on RLdata10000 , config file for today's run has HDFS hard coded paths. Both runs from  yesterday (7.9 Hrs) and today (7.7 Hrs) took around the same time , evaluation results are exactly same. 
[run.txt](https://github.com/cleanzr/dblink/files/3655077/run.txt)
[evaluation-results.txt](https://github.com/cleanzr/dblink/files/3655078/evaluation-results.txt)
[NewRLdata10000.conf.txt](https://github.com/cleanzr/dblink/files/3655079/NewRLdata10000.conf.txt)


cluster name and path has been disguised with xxxxxx in uploaded files.

**Comment by resteorts** [comment]
@YathishK This is great. Could you please now run ABSEmployee and report back please? Thank you so much for the update!

**Comment by YathishK** [comment]
1. Application is running on ABSEmployee.

2. We re-ran dblink on RLdata10000 with sample size of 10K  and  16 partitions (changed in "numlevels" to 4  in config file).  The job ran about 3 hours and failed with below error message 

_ERROR Utils: Uncaught exception in thread heartbeat-receiver-event-loop-thread_
java.lang.IllegalStateException: Attempted to access garbage collected accumulator 512002
 at org.apache.spark.util.AccumulatorContext$$anonfun$get$1.apply(AccumulatorV2.scala:265)
 at org.apache.spark.util.AccumulatorContext$$anonfun$get$1.apply(AccumulatorV2.scala:261)
 at scala.Option.map(Option.scala:146)

We are debugging the issue.

**Comment by YathishK** [comment]
"RLdata10000 with sample size of 10K and 16 partitions"  job log has below messages

19/09/26 17:29:42 ERROR AsyncEventQueue: Dropping event from queue eventLog. This likely means one of the listeners is too slow and cannot keep up with the rate at which tasks are being started by the scheduler.
19/09/26 17:29:42 WARN AsyncEventQueue: Dropped com.codahale.metrics.Counter@323691eb events from eventLog since Wed Dec 31 19:00:00 EST 1969.
19/09/26 17:30:42 WARN AsyncEventQueue: Dropped com.codahale.metrics.Counter@323691eb events from eventLog since Thu Sep 26 17:29:42 EDT 2019.
19/09/26 17:31:42 WARN AsyncEventQueue: Dropped com.codahale.metrics.Counter@323691eb events from eventLog since Thu Sep 26 17:30:42 EDT 2019.
19/09/26 17:32:42 WARN AsyncEventQueue: Dropped com.codahale.metrics.Counter@323691eb events from eventLog since Thu Sep 26 17:31:42 EDT 2019.
19/09/26 17:33:27 WARN KDTreePartitioner: Poor quality split (87.18775181305398%) at node 4.
19/09/26 17:33:27 WARN KDTreePartitioner: Poor quality split (85.26490066225165%) at node 14.
19/09/26 17:33:27 WARN KDTreePartitioner: Poor quality split (76.802284082798%) at node 9.
19/09/26 17:33:27 WARN KDTreePartitioner: Poor quality split (78.67590454195535%) at node 11.

**Comment by resteorts** [comment]
@YathishK As asked for many times, could you please send the specs of the machine please as otherwise it's not possible for further guidance to be given without this very important information.

**Comment by YathishK** [comment]
we are using below command :  spark-submit --master yarn --deploy-mode client --num-executors 69 --executor-cores 5 --executor-memory 6g  ( and other options such as log4 properties , .jar file and .conf file)

Driver node has 16GB  and worker nodes have 250GB memory

**Comment by resteorts** [comment]
@YathishK Could you please post the complete config and log files for the results your are reporting above? If you could always do this that would be very helpful in the event of an error such that we can see all relevant information.

**Comment by resteorts** [comment]
I have a few more basic questions that I just want to confirm with you for the RLdata10000 dataset. Let's focus on this one before working on the ABSEmployee or others. 

1. You are using 16 partitions, correct? 
2. How many worker threads are you using? 
3. Could you clarify the specs of you master node? Recall that we used  
an m4.large instance (4 vCore, 8 GiB memory, 32 GiB EBS storage)
4. Could you clarify the specs of you worker (slave) nodes? Recall that we 
Slave nodes: m5.xlarge instances (4 vCore, 16 GiB memory, 32 GiB EBS storage)
set the number of workers nodes so that we have one vCore per partition. 
So for RLdata10000, we used a master node + a SINGLE worker node.

If you could please give us the same type of information that you are using this would be helpful. If you could even repeat the same type of settings, this would be helpful too regarding just seeing how this works. Thank you and I hope this helps.

**Comment by YathishK** [comment]
Yes , first we will make RLdata10000 work for 16 partitions.  
1) You are using 16 partitions, correct?   Yes
2) How many worker threads are you using?   80
3)Could you clarify the specs of you master node?  ~250GB
4)Could you clarify the specs of you worker nodes?  ~250GB

Attached here the log file and config file for 16 partitions 
[D26RLdata10000_log.txt](https://github.com/cleanzr/dblink/files/3663176/D26RLdata10000_log.txt)
[P16RLdata10000.conf.txt](https://github.com/cleanzr/dblink/files/3663185/P16RLdata10000.conf.txt)

**Comment by ngmarchant** [comment]
Thanks for providing the log file.

The error and warning messages below relate to [SPARK-23867](https://issues.apache.org/jira/browse/SPARK-23867) which was fixed in Spark 2.3.1.
Note that we have only tested dblink with Spark 2.3.1.

> 19/09/26 18:03:51 ERROR AsyncEventQueue: Dropping event from queue eventLog. This likely means one of the listeners is too slow and cannot keep up with the rate at which tasks are being started by the scheduler.
> 19/09/26 18:03:51 WARN AsyncEventQueue: Dropped com.codahale.metrics.Counter@cb692aa events from eventLog since Wed Dec 31 19:00:00 EST 1969.
> 19/09/26 18:04:51 WARN AsyncEventQueue: Dropped com.codahale.metrics.Counter@cb692aa events from eventLog since Thu Sep 26 18:03:51 EDT 2019.
> 19/09/26 18:05:51 WARN AsyncEventQueue: Dropped com.codahale.metrics.Counter@cb692aa events from eventLog since Thu Sep 26 18:04:51 EDT 2019.
> 19/09/26 18:06:51 WARN AsyncEventQueue: Dropped com.codahale.metrics.Counter@cb692aa events from eventLog since Thu Sep 26 18:05:51 EDT 2019.

The messages below are just warning about partition size imbalance. The job will still run, but the load balancing may be suboptimal.

> 19/09/26 18:07:02 WARN KDTreePartitioner: Poor quality split (87.18775181305398%) at node 4.
> 19/09/26 18:07:02 WARN KDTreePartitioner: Poor quality split (85.26490066225165%) at node 14.
> 19/09/26 18:07:02 WARN KDTreePartitioner: Poor quality split (76.802284082798%) at node 9.
> 19/09/26 18:07:02 WARN KDTreePartitioner: Poor quality split (78.67590454195535%) at node 11.

The remaining lines indicate that the job was terminated before completion.

> Exception in thread "main" org.apache.spark.SparkException: Job aborted due to stage failure: Task serialization failed: java.lang.IllegalStateException: Cannot call methods on a stopped SparkContext.

**Comment by ngmarchant** [comment]
> ERROR Utils: Uncaught exception in thread heartbeat-receiver-event-loop-thread_
> java.lang.IllegalStateException: Attempted to access garbage collected accumulator 512002
> at org.apache.spark.util.AccumulatorContext$$anonfun$get$1.apply(AccumulatorV2.scala:265)
> at org.apache.spark.util.AccumulatorContext$$anonfun$get$1.apply(AccumulatorV2.scala:261)
> at scala.Option.map(Option.scala:146)

This error is probably an instance of [SPARK-22371](https://issues.apache.org/jira/plugins/servlet/mobile#issue/SPARK-22371) which was also fixed in Spark 2.3.1.
